lib/filterx/func-sdata.c

```diff
@@ -0,0 +1,313 @@
+
+  filterx_object_unref(fob_key);
+  filterx_object_unref(fob_value);
+
+  return res;
+}
+
+static gboolean
+_insert_while_same_sd_id(LogMessage *msg, guint8 index, guint8 num_sdata, guint8 *num_insertions,
+                         FilterXObject *inner_dict,
+                         const gchar *current_sd_id_start, gsize current_sd_id_len)
+{
+  const gchar *next_sd_id_start;
+  gsize next_sd_id_len;
+  const gchar *param_name_start;
+  gsize param_name_len;
+
+  *num_insertions = 0;
+
+  for (guint8 j = index + 1; j < num_sdata; j++)
+    {
+      gssize next_name_len;
+      const gchar *next_name = log_msg_get_value_name(msg->sdata[j], &next_name_len);
+      gssize next_value_len;
+      const gchar *next_value = log_msg_get_value(msg, msg->sdata[j], &next_value_len);
+
+      if (!_extract_sd_components(next_name, next_name_len, &next_sd_id_start, &next_sd_id_len, &param_name_start,
+                                  &param_name_len))
+        {
+          filterx_object_unref(inner_dict);
+          return FALSE;
+        }
+
+      if (_should_create_new_dict(current_sd_id_start, current_sd_id_len, next_sd_id_start, next_sd_id_len))
+        break;
+
+      if (!_insert_into_dict(inner_dict, param_name_start, param_name_len, next_value, next_value_len))
+        return FALSE;
+
+      (*num_insertions)++;
+    }
+
+  return TRUE;
+}
+
+static gboolean
+_generate(FilterXExprGenerator *s, FilterXObject *fillable)
+{
+  FilterXEvalContext *context = filterx_eval_get_context();
+  LogMessage *msg = context->msgs[0];
+  gboolean success = TRUE;
+
+  const gchar *current_sd_id_start;
+  gsize current_sd_id_len;
+  const gchar *param_name_start;
+  gsize param_name_len;
+
+  for (guint8 i = 0; i < msg->num_sdata;)
+    {
+      gssize name_len;
+      const gchar *name = log_msg_get_value_name(msg->sdata[i], &name_len);
+      gssize value_len;
+      const gchar *value = log_msg_get_value(msg, msg->sdata[i], &value_len);
+
+      if (!_extract_sd_components(name, name_len, &current_sd_id_start, &current_sd_id_len, &param_name_start,
+                                  &param_name_len))
+        {
+          success = FALSE;
+          goto error;
+        }
+
+      FilterXObject *inner_dict = filterx_object_create_dict(fillable);
+      success = _insert_into_dict(inner_dict, param_name_start, param_name_len, value, value_len);
+      if (!success)
+        goto error;
+
+      guint8 num_insertions;
+      success = _insert_while_same_sd_id(msg, i, msg->num_sdata, &num_insertions, inner_dict, current_sd_id_start,
+                                         current_sd_id_len);
+      if (!success)
+        goto error;
+      i += num_insertions + 1;
+
+      FilterXObject *sd_id_key = filterx_string_new(current_sd_id_start, current_sd_id_len);
+      filterx_object_set_subscript(fillable, sd_id_key, &inner_dict);
+      filterx_object_unref(inner_dict);
+      filterx_object_unref(sd_id_key);
+    }
+
+error:
```

Comment:
[nitpick]

It is strange to see that it is possible to go to the error returning part with `success = TRUE`, which does not return an error. I think we should either call this `exit`, or make `error` not depend on the value of `success`, or simply get rid of the gotos.

I would prefer the latter: usually we need the `goto`s, so it is easier to destroy local variables, but that is not needed here, so we ca just `return FALSE` instead of the lot of `goto`s. We might even be able to remove the `success` variable, too.

Reply:
You are absolutely right there, once I got rid of the things that would have required a cleanup this goto became obsolete, thanks!